### PR TITLE
Add callable support for `find` and `rfind` `Array` methods

### LINKED ISSUE #6889: Add more array methods: first_custom, last_custom, count_custom
### Describe the project you are working on

This is not unique to my project, but one project I am working on requires a lot of array traversals searching for elements that satisfy specific conditions. I feel there is a gap in Array functionality (because sort_custom already exists), and this would plug it,

### Describe the problem or limitation you are having in your project

There is no built-in methods to search for the first, last, and to count elements that satisfy custom conditions implemented in a Callable, the way you can filter or sort_custom.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add the following methods:
`find_custom(fn:Callable)` will return the index of the first element of an array where the callable fn will return true.
`rfind_custom(fn:Callable)` will return the index of the last element of an array where the callable fn will return true.
`count_custom(fn:Callable)` will return the count of all elements of the array where the callable fn will return true.

I attempted to name the functions to fit the naming style of the rest of the functions in the Array class. However, I do not like it- I would rather name it something like `find_by()`. But my gripes with the Array function naming conventions are beyond the scope of this proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a sketch implementation I have already created:
```cpp
int Array::find_custom(const Callable& p_callable, int p_from = 0) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0 || size() == 0) {
		return -1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'first_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0) {
		// Relative offset from the end
		p_from = _p->array.size() + p_from;
	}
	if (p_from < 0 || p_from >= _p->array.size()) {
		// Limit to array boundaries
		p_from = _p->array.size() - 1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i >= 0; i--) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'last_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::count_custom(const Callable& p_callable) const {
	int count = 0;
	const Variant* argptrs[1];

	for (int i = 0; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(0, "Error calling method from 'count_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Increase count if passes
		if (result.operator bool()) {
			count++;
		}
	}

	return count;
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, one can simply create a loop to accomplish the same task. But it must be done every time, and could be solved by patching this gap in GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

This adds core Array functionality.

### PR DESCRIPTION
**Pull request by SlashScreen**
I have implemented a number of methods for using callables with:
- `find` (`find_custom`)
- `rfind` (`rfind_custom`)
- `count` (`count_cusom`)

I implemented these in a previous PR, but I accidentally deleted it when cleaning up my github, so I am doing it again. Also, I am better at contributing to codebases now, and working with C++.

On the last PR, `count_custom` could just be `reduce`, and indeed that's what it uses internally, but I would argue that this is okay, as there is precedent for using simple wrapper methods to avoid confusion with less experienced programmers and work with their intuition (`has` relies on `find`, `append` is an alias for `push_back`). On the other hand, instead of `max_custom` and `min _custom`, you're instructed to use reduce instead, so there is precedence against, too.

Closes https://github.com/godotengine/godot-proposals/issues/6889

### CODE REVIEW

**Comment by SlashScreen** [comment]
Looks like I need to write tests. Also, I think my `count_custom` implementation using reduce is problematic. I can hand-roll it, or remove it entirely.

**Comment by SlashScreen** [comment]
Where are tests for `sort_custom` written?

File: core/variant/array.h
Comment (AThousandShips) [inline core/variant/array.h:157]:
```suggestion
	int rfind_custom(const Callable &p_callable, int p_from = -1) const;
```

File: core/variant/variant_call.cpp
Comment (AThousandShips) [inline core/variant/variant_call.cpp:2292]:
```suggestion
	bind_method(Array, find_custom, sarray("method"), varray(0));
```

File: core/variant/variant_call.cpp
Comment (AThousandShips) [inline core/variant/variant_call.cpp:2294]:
```suggestion
	bind_method(Array, rfind_custom, sarray("method"), varray(-1));
```

File: core/variant/array.cpp
Comment (AThousandShips) [inline core/variant/array.cpp:504]:
```suggestion
	Callable method = callable_mp_static(_count_reduce);
```

File: core/variant/array.cpp
Comment (AThousandShips) [inline core/variant/array.cpp:482]:
```suggestion
static int _count_reduce(int p_accum, const Variant &p_next, Callable &p_callable) {
```

**Comment by SlashScreen** [comment]
@AThousandShips resolved. Good catch. Also realized that `_count_reduce` wouldn't have actually done anything, so I fixed that.

A few remaining questions:
- Where should I put tests for these? 
- The CI build keeps failing on linux-mono, for some reason involving the doc tool. I don't totally understand the failure here.
- I briefly implemented a `take_while` method before realizing that it could be simplified to `find_custom` + `slice`. Should I add that back in, along with any more functional programming-style methods? I can see `skip` (creating a new array made of every `n`th element of the source) being useful, and there's also things like `zip` and `unzip`. Could be out of the scope of the PR, though.

File: core/variant/array.cpp
Comment (SlashScreen) [reply core/variant/array.cpp:482]: Ah, whoops. I'll resolve that now.

**Comment by AThousandShips** [comment]
> Where should I put tests for these?

Here `tests/core/variant/test_array.h`

Once you've fixed the callable argument issue we can see what the doc things are

File: core/variant/array.cpp
Comment (AThousandShips) [reply core/variant/array.cpp]: Any reason for a lack of reference for the `p_next`? It can reduce copying in some cases

**Comment by SlashScreen** [comment]
> > Where should I put tests for these?
> 
> Here `tests/core/variant/test_array.h`
> 
> Once you've fixed the callable argument issue we can see what the doc things are

I fixed the const issue. I was going to put the test in there, but I didn't see one for `filter`, so I wasn't sure if there was a special place I couldn't find. I'll add the tests now.

File: core/variant/variant_call.cpp
Comment (AThousandShips) [inline core/variant/variant_call.cpp:2292]:
```suggestion
	bind_method(Array, find_custom, sarray("method", "from"), varray(0));
```

File: core/variant/variant_call.cpp
Comment (AThousandShips) [inline core/variant/variant_call.cpp:2294]:
```suggestion
	bind_method(Array, rfind_custom, sarray("method", "from"), varray(-1));
```

**Review by AThousandShips — COMMENTED** [review]
This should fix CI

**Comment by SlashScreen** [comment]
I've added the tests, and fixed the bind method. Good catch. Makes sense that that failed.

**Comment by SlashScreen** [comment]
Sorry, I caught a typo.

**Comment by SlashScreen** [comment]
Tests run during CI, right?

**Comment by AThousandShips** [comment]
They do, and now even on release template builds for completeness

**Comment by SlashScreen** [comment]
Sorry about that. I know know how to escape characters in xml.

**Comment by AThousandShips** [comment]
Escaping XML characters is a secret art and even highly experienced developers here (like myself) struggle with them 🤣

**Comment by SlashScreen** [comment]
Who would need to approve this for it to get merged into main?

**Comment by AThousandShips** [comment]
The core team, but I'll give this my approval as well once I've tested it out a bit and gone over it, we'll hopefully release 4.3 soon and then this can get attention

**Comment by SlashScreen** [comment]
If I were to add other functions, like `take_while`, `skip`, `intersection`, `difference`, `deduplicate`, stuff like that, would that go in a PR or have to go through godot-proposals first?

**Comment by SlashScreen** [comment]
@AThousandShips Anything that needs to be changed?

**Comment by AThousandShips** [comment]
Will go over it but again the core team needs to review, and I'm not on the core team, will leave a final review once I've gone over it but I'll wait for the core team to give their opinion first

**Comment by SlashScreen** [comment]
Alrighty. I hope they get to it eventually...

**Comment by Mickeon** [comment]
> On the last PR, `count_custom` could just be `reduce`, and indeed that's what it uses internally, but I would argue that this is okay, as there is precedent for using simple wrapper methods to avoid confusion with less experienced programmers and work with their intuition (`has` relies on `find`, `append` is an alias for `push_back`).

I personally still don't see much of an use in `count_custom`, either. But I can also understand it may need to be there for completion sake. Not sure how strong the argument about less experienced programmers is, the given that passing Callables around is not exactly a beginner-friendly topic.

**Comment by SlashScreen** [comment]
@Mickeon 

That's true, `count_custom` is more or less just a wrapper over `reduce`. And I suppose passing around a `Callable` might not be easy for beginners. "Reduce" is the functional programming name for it, but I believe the name doesn't really communicate what it's capable of for people who haven't used it before, nor does the example. (I know because this was me last year.)

A compromise could be to add an example to use `reduce` to count, and adding a note to `count` that you can use `reduce` to accomplish the effect of `count_custom`.

**Comment by AThousandShips** [comment]
The only real benefit of `count_custom` is a marginal performance boost over using `for` in GDScript

**Comment by SlashScreen** [comment]
I see, so documentation it is.

**Comment by SlashScreen** [comment]
I'll get started on that tomorrow.

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:410]:
```suggestion
				print([1, 3, 4, 7].find_custom(func(num): return i % 2 == 0)) # prints "2"
```
We don't use static typing in the documentation as a rule, see [here](https://docs.godotengine.org/en/stable/contributing/documentation/docs_writing_guidelines.html#dynamic-vs-static-typing)

**Comment by AThousandShips** [comment]
Did you mean to restore the code for `count_custom`? You didn't bind it so either it should be bound or not implemented

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:640]:
```suggestion
				func is_even(val):
```

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:644]:
```suggestion
				    var arr = [1, 2, 3, 4, 5]
```

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:645]:
```suggestion
				    var even_count = arr.reduce(func(count, next): return count + 1 if is_even(next) else count, 0) # Increment count if it's even, else leaves count the same.
```

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:646]:
```suggestion
				    print(even_count) # Prints 2
```

**Review by AThousandShips — APPROVED** [review]
Approving on principle, though I'm not 100% sure these are needed, but I think they are useful even if they can be implemented in a straightforward way in scripting it's useful to avoid boilerplate 

But will defer to others for the final go-ahead or rejection, but I think they might be good to add

(You also need to remove the test for `count_custom`)

**Comment by AThousandShips** [comment]
In the future keep in mind you can apply suggestions as a batch, please do so in the future to not force many reruns of the CI

**Comment by AThousandShips** [comment]
Also later before merger the commit title should be changed to match the PR title to make it clearer

**Comment by SlashScreen** [comment]
Oops, I didn't know I could do that. Sorry,

File: doc/classes/Array.xml
Comment (AThousandShips) [inline doc/classes/Array.xml:410]:
```suggestion
				print([1, 3, 4, 7].find_custom(func(num): return i % 2 == 0)) # Prints 2
```
My bad missed this one

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:328]: Consistently referring to them as elements.
```suggestion
				To count how many elements in an array satisfy a condition, see [method reduce].
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:689]: This description is inconsistent with `find_custom`.
```suggestion
				Returns the index of the [b]last[/b] element of the array that causes [param method] to return [code]true[/code], or [code]-1[/code] if there are none. The search's start can be specified with [param from], continuing to the beginning of the array. This method is the reverse of [method find_custom].
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:638]:
```suggestion
				This method can also be used to count how many elements in an array satisfy a certain condition, similar to [method count]:
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:405]:
```suggestion
				Returns the index of the [b]first[/b] element in the array that causes [param method] to return [code]true[/code], or [code]-1[/code] if there are none. The search's start can be specified with [param from], continuing to the end of the array.
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:406]:
```suggestion
				[param method] is a callable that takes an element of the array, and returns a [bool].
```

**Review by Mickeon — COMMENTED** [review]
I have more suggestions but the PR is getting updated as I speak

Especially after https://github.com/godotengine/godot/pull/69451 has been in limbo for so long, I want to make sure this one is squeaky clean

File: doc/classes/Array.xml
Comment (Mickeon) [reply doc/classes/Array.xml]: I actually wanted to copy exactly how it is written in `any` for consistency, at least for the moment
```suggestion
				The [param method] should take one [Variant] parameter (the current array element) and return a [bool].
```

**Comment by SlashScreen** [comment]
Okay, added suggestions. Sorry about the CI Spam, lol

**Review by Mickeon — COMMENTED** [review]
I know it's correct wording but saying "satisfy a condition" is not something I see often in the class reference, I will think of something later.

File: doc/classes/Array.xml
Comment (Mickeon) [reply doc/classes/Array.xml]: I reiterate I would've preferred the above for consistency, but I will sleep soundly at night if the current description stays.

**Comment by SlashScreen** [comment]
if I were to build on this and add more array functions in a future PR, would that have to go through proposals first?

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:645]: The line is very, very long. I would suggest moving the comment above, and potentially making it more meaningful. Code examples should speak for themselves, too.
```suggestion
				   # Increments count if the number is even. Otherwise, leaves count unchanged.
				    var even_count = arr.reduce(func(count, number): return count + 1 if is_even(number) else count, 0)
```

Also, I'm not exactly sure about this right now, but I believe you can omit `0` here (like one of the snippets above), which would make the example less confusing.

```suggestion
				   # Increments count if the number is even. Otherwise, leaves count unchanged.
				    var even_count = arr.reduce(func(count, number): return count + 1 if is_even(number) else count)
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:641]: No need to be vague I suppose
```suggestion
				func is_even(number):
				    return number % 2 == 0
```

File: doc/classes/Array.xml
Comment (Mickeon) [inline doc/classes/Array.xml:410]: I would suggest unwrapping this singular line into 2 methods, like the `is_even` example or other methods that accept a custom callable.

**Comment by AThousandShips** [comment]
Yes, as they'd be different ones from these

File: doc/classes/Array.xml
Comment (SlashScreen) [reply doc/classes/Array.xml:645]: The default argument is `nil`, so wouldn't `nil + 1` cause an error?

Comment (SlashScreen) [reply doc/classes/Array.xml:645]: I have added it anyway, though

Comment (SlashScreen) [reply doc/classes/Array.xml]: @AThousandShips What does the thumbs up mean in this case

Comment (AThousandShips) [reply doc/classes/Array.xml]: Agreement, that it will be called with `null`

Comment (AThousandShips) [reply doc/classes/Array.xml]: It won't cause a bug with `null`, but it *will* use the first element of the array as `count` instead of passing `0` and the first element

Comment (SlashScreen) [reply doc/classes/Array.xml]: Okay, yeah, that's an issue.

**Comment by SlashScreen** [comment]
Alrighty, are we good to go?

**Comment by SlashScreen** [comment]
@Mickeon anything else that needs to be changed?

**Comment by AThousandShips** [comment]
Needs a review by the core team and preferably also the GDScript team before this is ready to merge

**Comment by SlashScreen** [comment]
Oh, I thought he was the core team... alright.

**Comment by AThousandShips** [comment]
Doc team, if you look at the top right you'll see that review requests are still pending there for teams, will go away when they review

**Comment by SlashScreen** [comment]
Oooh, okay, I see. None of my PRs adding features have gotten this far, so it's uncharted territory for me.

**Comment by SlashScreen** [comment]
Oooh, okay, I see. None of my PRs adding features have gotten this far, so it's uncharted territory for me

**Review by Nazarwadim — APPROVED** [review]

File: core/variant/array.cpp
Comment (KoBeWi) [inline core/variant/array.cpp:379]: I know this is copied from `find()`, but the `size()` check is the same as the condition at the top. You could merge these 2 conditions.

File: core/variant/array.cpp
Comment (KoBeWi) [inline core/variant/array.cpp:386]:
```suggestion
		const Variant &val = _p->array[i];
```

File: core/variant/array.cpp
Comment (KoBeWi) [inline core/variant/array.cpp:397]: tbh `ret` is unnecessary. You could just do `return i;` here.

**Comment by KoBeWi** [comment]
I wonder if we even need two methods 🤔
Precedent aside, negative `p_from` could make the search backwards.

**Review by KoBeWi — APPROVED** [review]
The implementation looks fine overall.

**Comment by Mickeon** [comment]
> I wonder if we even need two methods 🤔

I feel like separating it in two methods keeps the intention clearer. Regardless, It's better to keep it as is for consistency with the existing methods.

**Comment by SlashScreen** [comment]
If I added these changes, would I have to go through the approval process again?...

**Comment by KoBeWi** [comment]
No.
Also this still needs approval from core team anyway. I implemented similar methods before, so I can tell whether the code is ok, but core team has final saying whether the feature should be added (although it's a formality at this point).

**Comment by SlashScreen** [comment]
The CI says there's a duplicate declaration of Array? What?

**Comment by KoBeWi** [comment]
You accidentally pushed some wrong file
image

**Comment by SlashScreen** [comment]
oops.

File: core/variant/array.cpp
Comment (Nazarwadim) [inline core/variant/array.cpp:448]:
```suggestion
		if (unlikely(ce.error != Callable::CallError::CALL_OK)) {
```

File: core/variant/array.cpp
Comment (Nazarwadim) [inline core/variant/array.cpp:387]:
```suggestion
		if (unlikely(ce.error != Callable::CallError::CALL_OK)) {
```

**Comment by SlashScreen** [comment]
@Nazarwadim done.

File: tests/core/variant/test_array.h
Comment (dalexeev) [inline tests/core/variant/test_array.h:645]:
```suggestion
	int index = a1.find_custom(callable_mp_static(_find_custom_callable));
	CHECK_EQ(index, 2);
```

File: tests/core/variant/test_array.h
Comment (dalexeev) [inline tests/core/variant/test_array.h:652]:
```suggestion
	int index = a1.rfind_custom(callable_mp_static(_find_custom_callable));
	CHECK_EQ(index, 4);
```

**Comment by dalexeev** [comment]
To me `count_custom()` makes sense. Of course, `reduce()` can be used instead, but it is less convenient than a unary predicate. `reduce()` is a general function, and `count_custom()` is a common special case, in my opinion.

**Comment by SlashScreen** [comment]
> To me `count_custom()` makes sense. Of course, `reduce()` can be used instead, but it is less convenient than a unary predicate. `reduce()` is a general function, and `count_custom()` is a common special case, in my opinion.

I agree, but the consensus seemed to be to remove it.

**Comment by SlashScreen** [comment]
@Nazarwadim @dalexeev done.

**Review by akien-mga — APPROVED** [review]
This seems to have significant demand as there's been already at least 3 PRs attempting to implement this, and the linked proposal is somewhat consensual.

The proposed API seems fine to me. I didn't review implementation details but others did, so I think this is good to go.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
